### ISSUE TITLE
Deprecation Error with Price Modifiers file

### ISSUE THREAD

**Issue by ttmwebdesign** [CartThrob/support#21]
`Deprecated
Automatic conversion of false to array is deprecated
user/addons/cartthrob/ft.cartthrob_price_modifiers_configurator.php, line 257

Severity: E_DEPRECATED`

The site uses price modifiers. Well, there's two sites actually, both use price modifiers and both have the same deprecation error.
I just upgraded things, but this is all still local while I work through upgrading everything else.

Oh -- the other thing, is this error doesn't stick around. After a couple of reloads it disappears, which is weird, but not the first time I've seen that sort of thing happen with deprecation errors.

CT9.0.1 & EE7.5.22, PHP8.2

**Comment by ttmwebdesign** [comment]
Here's a similar error in the actual entry with the modifiers on it:

`Deprecated
Creation of dynamic property Cartthrob_price_modifiers_configurator_ft::$buttons_temp is deprecated
user/addons/cartthrob/ft.cartthrob_price_modifiers_configurator.php, line 379

Severity: E_DEPRECATED`

**Comment by ttmwebdesign** [comment]
I need to escalate this as a site-breaking issue now, not just a deprecation error.
I am getting this after selecting a couple of options and trying to add them to the cart:

<code>
TypeError Caught
base64_decode(): Argument #1 ($string) must be of type string, array given
ft.cartthrob_price_modifiers_configurator.php:117
</code>

Fortunately this is on my test site not the live site as we've not moved everything to the new PHP or softwares yet, but this also means I cannot proceed with doing that with the client's live site.

I can provide a link to the site in question, but I'd prefer to do this via DM in Slack vs. over the public GitHub issues.

Some additional information on this particular product as I compare it to CT5 - which works... I noticed on the front end of the live site that it has an option group label _with a # sign_ - but my test site is missing this on the front end. It *IS* present in the option group label field in the entry. I removed the # sign and retried the add-to-cart, and it worked without error. I added the # sign back, and it's not broken anymore - but it's also not displaying the full string from that label field on the front end, either.

I found another product with the hash sign in that label field (and also displaying it on the front end, so this is inconsistent and merits further investigation by me), but it is broken like the original was, so I am going to leave it alone in case it's needed for 'show and tell' as it were.

**Comment by robinsowell** [comment]
[issue-21.zip](https://github.com/user-attachments/files/27838548/issue-21.zip)

Let's see if this fixes it- you'll need to backup and replace

/addons/cartthrob/ft.cartthrob_price_modifiers_configurator.php
/user/addons/cartthrob/src/Controllers/Cp/Routes/Actions/ConfiguratorAjax.php

**Comment by ttmwebdesign** [comment]
That has solve the issue for both the deprecation and the more serious error. Thanks!

Will this be released in a future patch in the add-on store also?

**Comment by ttmwebdesign** [comment]
@robinsowell I am seeing a new (maybe? I didn't try this before to know if it was an issue before the patch) error, when trying to modify the configurator in an entry -- I am getting a 404 for a JS file in the CP:

404: Item does not exist
Sorry, we could not find the item you are trying to access in the system.
URL: /cp/addons/settings/cartthrob/configurator_ajax

The closest thing to this file that I can find is in themes/user/cartthrob_option_configurator/js/optionConfigurator_ajax.js.

Is that the one it is asking for and there's a typo somewhere?

It is causing a browser 'alert' style error message when I try to add or remove an option (see screenshot to see what area I'm referring to).

<img width="1226" height="310" alt="Image" src="https://github.com/user-attachments/assets/cbf745ff-7eee-4889-be24-8d137faea364" />

Thanks.

Additional information -- I was contemplating reverting CT to version seven or eight for this site, but I tested both two other sites we have that doesn't use these types of modifiers, and it's happening in them, too. The alert says "Request failed: error", with the same 404s in the console. How on earth does this sort of thing happen for so long? Are there really that few people using that kind of modifier (CartThrob Price Modifiers Configurator)?

**Comment by ttmwebdesign** [comment]
So, I leaned heavily on CoPilot for this, but I seem to have found a fix - as in the errors have stopped, and it's pulling the correct URL now - but I want you folks to look it over and make sure it's valid, e.g. safe, since I know next to nothing about PHP, and only slightly more about JS. See the zip file I attached with the two changed files.

[ct patch.zip](https://github.com/user-attachments/files/28125982/ct.patch.zip)

**Comment by ttmwebdesign** [comment]
There was another issue that the client brought up that exists in CT5 that he's just "been working around" all this time and it's still present in CT9 also - it's semi-related to this whole issue as it's all for the price modifiers configurator field... In the case of this latest issue, if a new row is added or deleted, the rows in the table below it, which could have information in them - SKU, prices, etc. - get wiped out completely.

It does this silently without errors.

Again with CoPilot, we pinned down what was happening - first, the two tables have the same ID, which is whatever the configurator table field is. Then there's something in the JS (same file as previous) that was targeting empty input, which lead to the data being deleted as it was being replaced with nothing. I've copied all of the back and forth between me and CoPilot if you need to read through it - just let me know.

Now, there are three files in total that patches all of the issues I posted after the file from Robin that fixed the first thing. I am attaching it here for reviewing. It also sorts out an async race issue that was happening with the initial patch CoPilot came up with, where, while adding new rows as no longer deleting the second table, deleting the new rows _didn't_ delete the equivalent row from the second table until another row was deleted. 

So, now the fixes are in here for: 
1. the invalid URL; whether or not inserting the CP Token into the middle of the URL was contributing to the errors, it's now adding those to the end no matter what, along with sorting out the issues caused by the extra 'cp'... Thus fixing the browser alert.
2. The disappearing table on row-add is addressed. New rows can be added or existing ones deleted and the second table reacts appropriately.

[ct patch v2.zip](https://github.com/user-attachments/files/28237029/ct.patch.v2.zip)

**Comment by robinsowell** [comment]
Ouch- that was indeed messy.  I've got a pr in for a fix.  I haven't compared with yours, but I'm sure it's likely the same basic approach.  I'd share the patch, but it's rolled in with a number of other small changes so ends up at 16 files.  I know we're shooting for a release soon, so I'll hold off if you're currently good with it.  But if not, yell and I can get one up.

**Comment by ttmwebdesign** [comment]
I wanted to come back to this to provide an update. After I used both the patch previously provided and the one I cobbled together with CoPilot, I did move ahead with upgrading the client's two sites (both use CartThrob and had the same issues). 

All is well, except I discovered another area where the URL that CT is generating in the links is bad in the same way as my previous notes. It's putting the CP token into the middle of the URL and it's 404'ing as a result. In this instance, it is when I am on the Order Manager page, and try to click on the link of one of the graphed items. Since the order manager is a separate add-on, I am guessing it has similar files like CT had, that will need to be adjusted. I will look for those and update them on my client's installs, and report back if they appear to be any different than what I found in CT itself.

Maybe this is something that was already picked up in the other changes you mentioned, but I wanted to bring it up just in case it wasn't.

**Comment by ttmwebdesign** [comment]
@robinsowell 
Line 57 of /addons/cartthrob_order_manager/views/om_reports_home.php ...

I've changed THIS:
```window.location.href = (EE.BASE + '/addons/settings/cartthrob_order_manager').replace("?S=0", "?").replace(/(S=[\w\d]+)?&D=cp(.*?)$/, "$2&$1") + '&' + value;```

to THIS:
```window.location.href = '<?= ee('CP/URL')->make('addons/settings/cartthrob_order_manager')->compile() ?>' + '&' + value;```
